From 5ed1dcd1cc3adf26c591eefc161495a4c313d15d Mon Sep 17 00:00:00 2001
From: Lars Schmertmann <Lars.Schmertmann@governikus.de>
Date: Thu, 13 Mar 2025 12:55:05 +0100
Subject: Fix usage of API functions on Windows 1607 (Build 14393)

* Windows Server 2016, Windows 10 LTSB 2016 and Windows 10 version 1607:
  SetThreadDescription is only available by Run Time Dynamic Linking in
  KernelBase.dll. See [1].
* According to [2] UiaRaiseNotificationEvent should be available on
  Windows Server 2016 but in fact it is not.

[1] https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setthreaddescription
[2] https://learn.microsoft.com/de-de/windows/win32/api/uiautomationcoreapi/nf-uiautomationcoreapi-uiaraisenotificationevent

Fixes: QTBUG-134075
Pick-to: 6.9.0 6.8 6.8.3
Change-Id: I3c4c733a4112a72b75f91f017a278dff2454e100
Reviewed-by: Thiago Macieira <thiago.macieira@intel.com>
(cherry picked from commit a9e251332cbec0f64c9c085349836af7276c55d3)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/thread/qthread_win.cpp | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

(limited to 'src/corelib/thread')

diff --git a/src/corelib/thread/qthread_win.cpp b/src/corelib/thread/qthread_win.cpp
index 3b9b8871e22..77868b13a04 100644
--- a/src/corelib/thread/qthread_win.cpp
+++ b/src/corelib/thread/qthread_win.cpp
@@ -165,7 +165,21 @@ unsigned int __stdcall QT_ENSURE_STACK_ALIGNED_FOR_SSE QThreadPrivate::start(voi
     QString threadName = std::exchange(thr->d_func()->objectName, {});
     if (Q_LIKELY(threadName.isEmpty()))
         threadName = QString::fromUtf8(thr->metaObject()->className());
+#ifndef QT_WIN_SERVER_2016_COMPAT
     SetThreadDescription(GetCurrentThread(), reinterpret_cast<const wchar_t *>(threadName.utf16()));
+#else
+    HMODULE kernelbase = GetModuleHandleW(L"kernelbase.dll");
+    if (kernelbase != NULL) {
+        typedef HRESULT (WINAPI *DESCFUNC)(HANDLE, PCWSTR);
+
+        DESCFUNC setThreadDescription =
+            (DESCFUNC)GetProcAddress(kernelbase, "SetThreadDescription");
+        if (setThreadDescription != NULL) {
+            setThreadDescription(GetCurrentThread(),
+                                 reinterpret_cast<const wchar_t *>(threadName.utf16()));
+        }
+    }
+#endif
 
     emit thr->started(QThread::QPrivateSignal());
     QThread::setTerminationEnabled(true);
-- 
cgit v1.2.3

